lib/capistrano/net_storage/utils.rb

```diff
@@ -18,28 +18,34 @@ def config
       # You can provide either of +dest_dir+ or +dest_path+, or files are copied to same pathes
       def upload_files(files, dest_dir: nil, dest_path: nil)
         c = config
-        files.each do |src|
-          basename = File.basename(src)
-          dest = dest_path || begin
-            dir = dest_dir || File.dirname(src)
-            File.join(dir, basename)
-          end
-
-          if c.upload_files_by_rsync?
-            Parallel.each(c.servers, in_threads: c.max_parallels) do |host|
+        if c.upload_files_by_rsync?
+          on c.servers, in: :groups, limit: c.max_parallels do |host|
+            files.each do |src|
+              dest = build_dest(src, dest_dir: dest_dir, dest_path: dest_path)
               ssh = build_ssh_command(host)
               run_locally do
                 execute :rsync, "-az --rsh='#{ssh}' #{src} #{host}:#{dest}"
               end
             end
-          else
-            on c.servers, in: :groups, limit: c.max_parallels do
+          end
+        else
+          on c.servers, in: :groups, limit: c.max_parallels do
+            files.each do |src|
+              dest = build_dest(src, dest_dir: dest_dir, dest_path: dest_path)
               upload! src, dest
             end
           end
         end
       end
 
+      def build_dest(src, dest_dir: nil, dest_path: nil)
```

Comment:
Because this is an utility module which is used by other modules, I don't like to add this method in this module.

Reply:
Fuumm
Where place is better to add this module?
`C::NetStorage::Config.build_dest` ?

Reply:
My suggestion is to un-refactor the loop; that is, to revert partially the previous commit 18152fb9393b4bb29eae0d92541bd3973f8f2c07 .
Or you can use nameless method of Proc.

I don't think we need this method to reuse anywhere.

Reply:
I think we can remove this code totally considering we provide files as source to `rsync` as following:

* If the destination is a directory, `rsync` will sync the file on the destination with the same `basename` as the original one.
* If the destination is a file, `rsync` will sync the file content.

```ruby
def upload_files(files, dest_dir)
  files.each do |file|
    execute :rsync, "-az --rsh='#{ssh}' #{src} #{host}:#{dest}"
  end
end
```

Reply:
@aeroastro 

I think you are correct and your opinion is understandable. :+1: